# Exercise - Assess Data Quality and Structure Visually

In this exercise, we will be looking at synthetic phase two clinical trial dataset of 350 patients for a new innovative oral insulin called **Auralin** - a proprietary capsule that can solve a stomach lining problem. 

**Note:** Auralin and Novodra are not real insulin products. This clinical trial data was fabricated for the sake of this course. When assessing this data, the issues that you'll detect (and later clean) are meant to simulate real-world data quality and tidiness issues with the capability to **impact quality of care, patient registration, and revenue**. The datasets, `patients` and `treatments`, were constructed with the consultation of real doctors to ensure plausibility.

In [1]:
#DO NOT MODIFY - imports and loaing data
import pandas as pd

patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')

## 1. Identify data quality issues in the patients data

The `patients` dataframe contains the following variables:

- **patient_id**: the unique identifier for each patient
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year).
> The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. 
> BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

### 1.1 Display dataset

Display the first 10 rows of the patients data, and visually inspect the data, using directed and non-directed visual assessment as you see necessary. 

Identify **four** instances of problematic data points corresponding to **accuracy** and **validity** in the patients dataset.

**Hint**: take a look at the `given_name`, `zip_code`, and `height` columns. 

In [3]:
#FILL IN - inspect the first 10 rows of the patients data

patients.head(10)

patient_id assigned_sex given_name     surname                     address  \
0           1       female        Zoe     Wellish        576 Brown Bear Drive   
1           2       female     Pamela        Hill   2370 University Hill Road   
2           3         male        Jae      Debord       1493 Poling Farm Road   
3           4         male       Liêm        Phan         2335 Webster Street   
4           5         male        Tim     Neudorf        1428 Turkey Pen Lane   
5           6         male     Rafael       Costa          1140 Willis Avenue   
6           7       female       Mary       Adams            3145 Sheila Lane   
7           8       female     Xiuxiu       Chang  2687 Black Oak Hollow Road   
8           9         male      Dsvid  Gustafsson          1790 Nutter Street   
9          10       female     Sophie     Cabrera           3303 Anmoore Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   
5      Daytona Beach     Florida   32114.0  United States   
6            Burbank          NV   84728.0  United States   
7        Morgan Hill          CA   95037.0  United States   
8        Kansas City          MO   64105.0  United States   
9           New York    New York   10011.0  United States   

                                         contact   birthdate  weight  height  \
0           951-719-9170ZoeWellish@superrito.com   7/10/1976   121.7      66   
1          PamelaSHill@cuvox.de+1 (217) 569-3204    4/3/1967   118.8      66   
2               402-363-6804JaeMDebord@gustr.com   2/19/1980   177.8      71   
3     PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   220.9      70   
4                334-515-7487TimNeudorf@cuvox.de   2/18/1928   192.3      27   
5       386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   183.9      70   
6              775-533-5933MaryBAdams@einrot.com  11/19/1969   146.3      65   
7           XiuxiuChang@einrot.com1 408 778 3236   8/13/1958   158.0      60   
8        816-265-9578DavidGustafsson@armyspy.com    3/6/1937   163.9      66   
9  SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7      64   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1  
5  26.4  
6  24.3  
7  30.9  
8  26.5  
9  33.4

In [4]:
#FILL IN - More code if needed
patients.nunique()

patient_id      503
assigned_sex      2
given_name      470
surname         466
address         483
city            349
state            54
zip_code        430
country           1
contact         483
birthdate       493
weight          347
height           22
bmi             183
dtype: int64

In [20]:
# Select specific columns and display the first 10 rows
patients[['given_name', 'zip_code', 'height']].head(10)
#patients['given_name'].unique()
patients['zip_code'].unique()
#patients['height'].unique()
# Filter rows where 'zip_code' is None and display the first 10
patients[patients['zip_code'].isna()].head(10)


patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4

In [38]:
# Unique 
patients[patients['surname']=='Doe']

patients['contact'].unique()

array(['951-719-9170ZoeWellish@superrito.com',
       'PamelaSHill@cuvox.de+1 (217) 569-3204',
       '402-363-6804JaeMDebord@gustr.com',
       'PhanBaLiem@jourrapide.com+1 (732) 636-8246',
       '334-515-7487TimNeudorf@cuvox.de',
       '386-334-5237RafaelCardosoCosta@gustr.com',
       '775-533-5933MaryBAdams@einrot.com',
       'XiuxiuChang@einrot.com1 408 778 3236',
       '816-265-9578DavidGustafsson@armyspy.com',
       'SophieCabreraIbarra@teleworm.us1 718 795 9124',
       '906-478-8949SandyGunnarsson@dayrep.com',
       'Abdul-NurMummarIsa@rhyta.com1 931 207 0839',
       'OmeokachieIbeamaka@einrot.com434-509-2614',
       'AnenechiChidi@armyspy.com+1 (205) 417-8095',
       'AsiaWozniak@rhyta.com918-712-3469',
       '276-225-1955SrenFLund@gustr.com',
       'LieuThiThuTam@dayrep.com1 559 765 7836',
       'RoxanneAndreyeva@armyspy.com678-829-8578',
       '406-775-2696WilliamVOates@armyspy.com',
       'ZakKelly@rhyta.com1 530 532 8397',
       'SofiaTKarlsen@teleworm.us1 

### 1.2 Identify issues

FILL IN the identified issues. For each issue, denote the row and column where the issue takes place:

- Issue 1 (Accuracy): Row ..., column ...  Address is missing in a number of columns
- Issue 2 (Accuracy): Row ... email and phone numbers are stored in one variable
- Issue 3 (Validity): Row ..., 215 ...  There are duplicates for John Doe in terms of address and patients id.
- Issue 4 (Validity): Row ..., email is inconsistant with phone numbers stored in the begining or the end of the string

## 2. Identify data quality and structural issues in the treatments data

Let's take a look at brief context around the `treatments` data:

350 patients participated in this clinical trial. None of the patients were using **Novodra** (a popular injectable insulin) or **Auralin** (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.
> HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash `-`) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash `-`). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment.

- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment.
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start - hba1c_end`.

### 2.1 Display data

Display the first 10 rows of the `treatments` data, and visually inspect the data. Identify 3 instances of problematic data quality issues corresponding to **completeness** and **consistency** in the dataset.

Hints: 
- For the first completeness issue, recall that the size of each treatment arm was 175 patients in each for the Auralin and Novodra arms. How many records should we have in the dataset, and how many rows does the treatments data currently contain?
- For the second completeness issue, take a look at the `hba1c_change` variable. Is there anything noticable about the values for this column? What could we do to mitigate this during cleaning?
- For the consistency issue, take a look at the `given_name` and `surname` variables. How are these different from what we saw in the `patients` dataframe?

In [39]:
#FILL IN - inspect the first 10 rows of the treatments data
treatments.head(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   
5    jasmine       sykes          -  42u - 44u         7.56       7.18   
6     sophia      haugen  37u - 42u          -         7.65       7.27   
7      eddie      archer  31u - 38u          -         7.89       7.55   
8      saber      ménard          -  54u - 54u         8.08       7.70   
9       asia     woźniak  30u - 36u          -         7.76       7.37   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32  
5          0.38  
6          0.38  
7          0.34  
8           NaN  
9           NaN

In [42]:
#FILL IN - calculate the number of rows in your data
treatments.shape

(280, 7)

In [44]:
#Fill in - More code if needed
treatments.isna().

given_name  surname  auralin  novodra  hba1c_start  hba1c_end  \
0         False    False    False    False        False      False   
1         False    False    False    False        False      False   
2         False    False    False    False        False      False   
3         False    False    False    False        False      False   
4         False    False    False    False        False      False   
..          ...      ...      ...      ...          ...        ...   
275       False    False    False    False        False      False   
276       False    False    False    False        False      False   
277       False    False    False    False        False      False   
278       False    False    False    False        False      False   
279       False    False    False    False        False      False   

     hba1c_change  
0            True  
1           False  
2            True  
3           False  
4           False  
..            ...  
275         False  
276          True  
277         False  
278         False  
279         False  

[280 rows x 7 columns]

### 2.2. Identifiy issues

FILL IN the identified issues. For each issue, denote the row and column where the issue takes place:

- Issue 1 (Completeness): Row ..., column ....  The issue is...
- Issue 2 (Completeness): Row ..., column ....  ...
- Issue 3 (Consistency): Row ..., column ....  ...

### 2.3 Data structural issue 

Identify one data structural issue in the dataset - take a look at the `auralin` and `novodra` columns. Why are these variables causing a messy dataset (what rule of tidiness do they violate?), and how should they be structured instead?

*FILL IN*

- Issue (Structural): ...